src/meta/src/hummock/compaction/picker/space_reclaim_compaction_picker.rs

```diff
@@ -0,0 +1,301 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+
+use risingwave_pb::hummock::hummock_version::Levels;
+use risingwave_pb::hummock::{InputLevel, SstableInfo};
+
+use crate::hummock::compaction::{CompactionInput, CompactionPicker, LocalPickerStatistic};
+use crate::hummock::level_handler::LevelHandler;
+
+pub struct SpaceReclaimCompactionPicker {
+    // config
+    pub max_space_reclaim_bytes: u64,
+    pub all_table_ids: HashSet<u32>,
+
+    // state
+    pub last_select_index: usize,
+}
+
+impl SpaceReclaimCompactionPicker {
+    pub fn new(max_space_reclaim_bytes: u64, all_table_ids: HashSet<u32>) -> Self {
+        Self {
+            max_space_reclaim_bytes,
+            last_select_index: 0,
+            all_table_ids,
+        }
+    }
+
+    fn filter(&self, sst: &SstableInfo) -> bool {
+        let table_id_in_sst = sst.table_ids.iter().cloned().collect::<HashSet<u32>>();
+        // it means all the table exist , so we not need to pick this sst
+        table_id_in_sst
+            .iter()
+            .all(|id| self.all_table_ids.contains(id))
+    }
+}
+
+impl CompactionPicker for SpaceReclaimCompactionPicker {
+    fn pick_compaction(
+        &mut self,
+        levels: &Levels,
+        level_handlers: &[LevelHandler],
+        _stats: &mut LocalPickerStatistic,
+    ) -> Option<CompactionInput> {
+        assert!(!levels.levels.is_empty());
+        let reclaimed_level = levels.levels.last().unwrap();
+        let mut select_input_ssts = vec![];
+        let level_handler = &level_handlers[reclaimed_level.level_idx as usize];
+
+        if self.last_select_index >= reclaimed_level.table_infos.len() {
```

Comment:
Why do we need to maintain the `last_select_index`? If the selected SSTs are marked pending after `pick_compaction`, I think these SSTs will be filtered out by `is_pending_compact` check anyway so `last_select_index` seems unneccessary?

Reply:
Also, since the table_infos list can change across different compaction runs, maintaining `last_select_index` looks inaccurate to me.

Reply:
Indeed pending will ensure that we don't select the same file if there are multiple triggers.

Suppose a situation: there is only one trigger path, the picker selects 5 files from index 0 in order, but the task is canceled due to some restrictions (such as delete_ratio too low), then continue to select these 5 files.

However, if we do not have the logic of active cancellation, will there be no above-mentioned problems? last_select_index is only used to ensure that the process of selecting files can continue to move forward (why this pr make picker stateful)